apps/drupal/src/Plugin/Preprocessor/PreprocessorManager.php

```diff
@@ -0,0 +1,432 @@
+<?php
+
+namespace Drupal\particle\Plugin\Preprocessor;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\particle\Tools\ParticleTools;
+
+/**
+ * Provides a theme preprocessor manager for Class based theme preprocessors.
+ *
+ * Drupal core does not provide built-in plugin support for theme files.
+ * The plugin discovery might before the front end theme is not active.
+ *
+ */
+class PreprocessorManager {
+
+  /**
+   * Cached preprocessor class map.
+   *
+   * @var array
+   */
+  protected static $preprocessorClassMap;
+
+  /**
+   * Appended suggestions.
+   *
+   * @var array
+   */
+  protected static $allSuggestions = [];
+
+  /**
+   * Cached instances.
+   *
+   * @var array
+   */
+  protected static $preprocessorInstances = [];
+
+  /**
+   * Particle tools instance for all preprocessors.
+   *
+   * @var Drupal\particle\Tools\ParticleTools
+   */
+  protected static $tools;
+
+  /**
+   * Returns the tools object.
+   *
+   * @return Drupal\particle\Tools\ParticleTools
+   *   The tools object.
+   */
+  public static function getTools() {
+    if (!isset(static::$tools)) {
+      static::$tools = new ParticleTools();
+    }
+
+    return static::$tools;
+  }
+
+  /**
+   * Returns the full path to the directory of the preprocess classes.
+   *
+   * @return string
+   *   The directory path.
+   */
+  public static function getPreprocessorsDirectory() {
+    return __DIR__ . '/Preprocessors';
+  }
+
+  /**
+   * Get the preprocessors from directory and subdirectories.
+   *
+   * @return array
+   *   An array of preprocessors from the file system.
+   */
+  public static function getPreprocessors() {
+    $preprocessors = [];
+    $directory = scandir(static::getPreprocessorsDirectory());
+    // Remove Base, '.' and '..' values and return directory names.
+    $types = array_slice($directory, 3);
+    foreach ($types as $type) {
+      $items = scandir(static::getPreprocessorsDirectory() . '/' . $type);
+      foreach ($items as $item) {
+        if (!is_dir($item)) {
+          $item = str_replace('.php', '', $item);
+          $preprocessors[$item] = $type;
+        }
+      }
+    }
+
+    return $preprocessors;
+  }
+
+  /**
+   * Get the preprocessor class map.
+   *
+   * @return array
+   *   An array with keys of preprocess name and values of full class path.
+   */
+  public static function getPreprocessorsClassMap() {
+    if (!isset(static::$preprocessorClassMap)) {
```

Comment:
@jryanconklin This cannot be a one-time cached variable. This uses static::$allSuggestions which may not be built yet or might be incomplete. Example: First call to this is for node.card and static::$allSuggestions is populated with suggestions for node in a card view mode. The next call is for a node in hero view mode (or even a different entity type - paragraph), then the static::$preprocessorClassMap would not get updated since it already was set which forces node.hero to fallback to the "default" class.